pageserver/src/tenant/mgr.rs

```diff
@@ -949,6 +949,19 @@ impl TenantManager {
                 (LocationMode::Attached(attach_conf), Some(TenantSlot::Attached(tenant))) => {
                     match attach_conf.generation.cmp(&tenant.generation) {
                         Ordering::Equal => {
+                            if attach_conf.attach_mode == AttachmentMode::Single {
+                                // We need to wait for LSN lease duration after transitioning into `AttachedSingle`
+                                // before doing any gc so that all previously granted lease will expire.
+                                let tenant = tenant.clone();
+                                tokio::spawn(async move {
+                                    let lsn_lease_length = tenant.get_lsn_lease_length();
+                                    let _ = tenant
+                                        .gc_block
+                                        .block_for(lsn_lease_length, &tenant.cancel)
+                                        .await;
+                                });
```

Comment:
This is quite the hack.

The `tokio::sync::Mutex` is used to replace the earlier "scary" communication using notify, not necessarily to block. I had not thought of this use-case ...

Reply:
unresolving due to tokio::spawn still existing, block_for still being used.

Reply:
Removed `block_for` in favor of recording blocked until deadline: 7d8bbddbb990ad34452889b3dc1df09c03f084f3